actors/evm/src/interpreter/instructions/lifecycle.rs

```diff
@@ -162,16 +162,18 @@ pub fn selfdestruct(
         return Err(StatusCode::StaticModeViolation);
     }
 
-    // Try to give funds to the beneficiary. We don't use the `delete_actor` syscall to do this
-    // because that won't auto-create the beneficiary (and will fail if, for some reason, we can't
-    // send them the funds).
-    //
-    // If we fail, we'll just burn the funds. Yes, this is what the EVM does.
+    // Try to give funds to the beneficiary. If this fails, we just keep them.
     if let Ok(addr) = EthAddress::from(beneficiary).try_into() {
         let balance = system.rt.current_balance();
         let _ = system.rt.send(&addr, METHOD_SEND, None, balance);
     }
-    // Now try to delete ourselves. If this fails, we abort execution.
-    system.rt.delete_actor(&BURNT_FUNDS_ACTOR_ADDR)?;
-    Ok(Output { outcome: Outcome::Delete, return_data: Bytes::new() })
+
+    // Now mark ourselves as deleted.
+    system.mark_selfdestructed();
+
+    // And "return".
+    //
+    // 1. In the constructor, this will set our code to "empty". This is correct.
```

Comment:
I guess another (theoretical-only) option would be to revert to Placeholder state, but it doesn't buy us much, and in fact it loses information, because we know that the actor sitting at this address can only ever be a smart contract.

Question: are we confident that this interacts correctly with all the method translation/dispatch shenanigans at the JSON-RPC layer and CALL opcodes?

Reply:
Unfortunately, we have no way to turn into an actual placeholder right now.

Reply:
> Question: are we confident that this interacts correctly with all the method translation/dispatch shenanigans at the JSON-RPC layer and CALL opcodes?

In terms of CALL opcodes, everything will behave the same. An "empty" contract is defined to just return without doing anything.

In terms of the JSON-RPC, this will behave like an empty contract (accepts calls) instead of like an embryo (rejects non-method-0 calls). That's technically "more correct" with respect to the EVM, but differs from existing behavior. See https://github.com/filecoin-project/ref-fvm/issues/1302.